File: 程序员面试金典第6版/面试题 08.07. Permutation I LCCI.cpp
// 该解法同样适用于有重复字符的情况
// 请注意 只有原始输入时字典序的时候 next_permutation才能生成全部的序列

class Solution {
public:
    vector<string> permutation(string S) {
        sort(S.begin(), S.end());
        vector<string> res;
        res.push_back(S);
        while(next_permutation(S.begin(), S.end())) {
            res.push_back(S);
        }
        return res;
    }
};

File: 程序员面试金典第6版/面试题 08.07. Permutation I LCCI.py
# 82% faster
# bfs
class Solution:
    def permutation(self, S: str) -> List[str]:
        res = []
        def generate(cur, remain):
            if not remain:
                res.append(cur)
                return
            for i, c in enumerate(remain):
                generate(cur+c, remain[:i]+remain[i+1:])
        generate("", S)
        return res